### DISCUSSION TITLE
Can i create custom localhost server?

### DISCUSSION THREAD

**Discussion by BrainRTP** [MrOkiDoki/BattleBit-Community-Server-API#86] [Q&A] [answered]
Hi, I want to try to create my own server to try to realize some idea (zombie mob). 
To do this I need to start the server locally and try to join it.
I understand that now there is no possibility to develop and test servers locally? 
By the example of other games, such as Minecraft, you can connect to localhost and check the server, whether it works correctly. 
I couldn't find any information on how to connect to my server when I looked at the documentation. 

Can you please tell me what I am missing?

**Comment by Axiomancer** [comment, marked as answer]
You cannot create a server localhost at the minute.

**Reply by BrainRTP** [reply]
So for development I need to run on some remote server (VPS/dedicated serve) with a public IP?

**Reply by Axiomancer** [reply]
Yes and that IP needs to be whitelisted by Oki to connect to the master server, without it you won't be able to do anything

We are petitioning to change this approach so servers can have staging environments

**Reply by BrainRTP** [reply]
Saw on discord that recruitment was already over. 
I take it there is no point in applying for IP whitelisting at this point?

**Reply by Bims-sh** [reply]
It's closed for now but look out for further updates from Oki

**Reply by Axiomancer** [reply]
I don't speak on behalf of BB or Oki so it may be worth speaking with him but to my knowledge, they are not onboarding any more community servers at this time 

@MrOkiDoki

**Reply by BrainRTP** [reply]
I have a small VPS (personal), I would like to try my hand at server writing, but I will only be able to write in my spare time. 
If there is a possibility to add a server to the whitelist - it would be great (just tell me where to go), but I can't say when the server will be ready because for me the development of this server will be something else than a hobby in my spare time. 
I'm not going to let other players in until I'm sure that my idea, which I will realize, is tested and ready to accept players.

All in all, it would be really cool to see soon the ability to connect to localhost for more convenient development bypassing the master server. I think it would make everyone's life easier, especially server developers :) 

@MrOkiDoki fyi

**Reply by Bims-sh** [reply]
@BrainRTP fyi, the game server currently only runs on Windows servers. There are linux options but they're not as stable as we'd want them to be 👍

**Reply by Axiomancer** [reply]
If you are really passionate Brain, you can look to join the Battlebit API Runner discord (https://discord.gg/FTkj9xUvHh) that modulates the creation of custom scripts, you could at least try your hand at that but knowing you cant reallllllly test without a server other than successful c# compiles. It'll also give you a chance to work with other server owners and see if you can contribute towards what they build

**Reply by BrainRTP** [reply]
@DasIschBims Yes, I realize it only works on Windows. 
But there is Docker :)
In general, docker will be enough for me to "touch" the server (since I can't do it locally). 

In general, this problem is solvable, after all, there is an API, and you can communicate via TCP in any programming language (but this is a topic for a separate discussion).

**Reply by Bims-sh** [reply]
@BrainRTP Yes indeed, I made an docker compose file, also like AxioInfinity said check out the modular API if you want to ^^
